Swift: Initial language guides documentation for Swift (#12940)
github/codeql: docs/codeql/codeql-language-guides/analyzing-data-flow-in-swift.rst

The following sections describe how to use the libraries for local data flow, global data flow, and taint tracking.
For a more general introduction to modeling data flow, see ":ref:`About data flow analysis <about-data-flow-analysis>`."

Local data flow
---------------

Local data flow tracks the flow of data within a single function. Local data flow is easier, faster, and more precise than global data flow. Before looking at more complex tracking, you should always consider local tracking because it is sufficient for many queries.

Using local data flow
~~~~~~~~~~~~~~~~~~~~~

You can use the local data flow library by importing the ``DataFlow`` module. The library uses the class ``Node`` to represent any element through which data can flow.
The ``Node`` class has a number of useful subclasses, such as ``ExprNode`` for expressions and ``ParameterNode`` for parameters. You can map between data flow nodes and expressions/control-flow nodes using the member predicates ``asExpr`` and ``getCfgNode``:

.. code-block:: ql

     class Node {
       /**
        * Gets this node's underlying expression, if any.
        */

>        Expr asExpr() { none() }

Author: Good idea.  I can see that `none()` could be confusing!

Using local taint tracking
~~~~~~~~~~~~~~~~~~~~~~~~~~

Local taint tracking extends local data flow to include flow steps where values are not preserved, such as string manipulation.
For example:

.. code-block:: swift

     temp = x
     y = temp + ", " + temp

If ``x`` is a tainted string then ``y`` is also tainted.

The local taint tracking library is in the module ``TaintTracking``.
Like local data flow, a predicate ``localTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo)`` holds if there is an immediate taint propagation edge from the node ``nodeFrom`` to the node ``nodeTo``.
You can apply the predicate recursively, by using the ``+`` and ``*`` operators, or you can use the predefined recursive predicate ``localTaint``.

For example, you can find taint propagation from an expression ``source`` to an expression ``sink`` in zero or more local steps:

.. code-block:: ql

     TaintTracking::localTaint(DataFlow::exprNode(source), DataFlow::exprNode(sink))

Examples of local data flow
~~~~~~~~~~~~~~~~~~~~~~~~~~~

This query finds the ``format`` argument passed into each call to ``String.init(format:_:)``:

.. code-block:: ql

    import swift

>     from CallExpr call, MethodDecl method

Author: Excellent point.

    from CallExpr call, MethodDecl method, Expr sourceExpr, Expr sinkExpr
    where
      call.getStaticTarget() = method and
      method.hasQualifiedName("String", "init(format:_:)") and
      sinkExpr = call.getArgument(0).getExpr() and
      DataFlow::localFlow(DataFlow::exprNode(sourceExpr), DataFlow::exprNode(sinkExpr))
    select sourceExpr, sinkExpr

We can vary the source, for example, making the source the parameter of a function rather than an expression. The following query finds where a parameter is used for the format:

.. code-block:: ql

    import swift
    import codeql.swift.dataflow.DataFlow

    from CallExpr call, MethodDecl method, ParamDecl sourceParam, Expr sinkExpr
    where
      call.getStaticTarget() = method and
      method.hasQualifiedName("String", "init(format:_:)") and
      sinkExpr = call.getArgument(0).getExpr() and
      DataFlow::localFlow(DataFlow::parameterNode(sourceParam), DataFlow::exprNode(sinkExpr))
    select sourceParam, sinkExpr

The following example finds calls to ``String.init(format:_:)`` where the format string is not a hard-coded string literal:

.. code-block:: ql

    import swift
    import codeql.swift.dataflow.DataFlow

>     from CallExpr call, MethodDecl method, Expr sinkExpr
>     where
>       call.getStaticTarget() = method and
>       method.hasQualifiedName("String", "init(format:_:)") and
>       sinkExpr = call.getArgument(0).getExpr() and
>       not exists(StringLiteralExpr sourceLiteral |
>         DataFlow::localFlow(DataFlow::exprNode(sourceLiteral), DataFlow::exprNode(sinkExpr))
>       )
>     select call, "Format argument to " + method.getName() + " isn't hard-coded."

Reviewer: This will have a FP when there's no dataflow node for `sinkExpr` (for example, if the expression isn't reachable because the control-flow library determined that the code was dead). We have observed this in C++, which is why we recently rewrote this example query when we updated the documentation with use-use flow (see https://codeql.github.com/docs/codeql-language-guides/analyzing-data-flow-in-cpp-new/#examples). Based on the way we wrote it for C++, I suggest something like this:

```suggestion
    from CallExpr call, MethodDecl method, DataFlow::Node sinkNode
    where
      call.getStaticTarget() = method and
      method.hasQualifiedName("String", "init(format:_:)") and
      sinkNode.asExpr() = call.getArgument(0).getExpr() and
      not exists(StringLiteralExpr sourceLiteral |
        DataFlow::localFlow(DataFlow::exprNode(sourceLiteral), sinkNode)
      )
    select call, "Format argument to " + method.getName() + " isn't hard-coded."
```

Feel free to modify this to use `DataFlow::exprNode` instead of `asExpr` if you think that is more pedagogically helpful.

Author: Thanks for this subtle correction, and yes I'm happy with `DataFlow::Node` and `asExpr` they're already discussed in the doc so should be familiar to the reader.

Author: I've created a merge conflict with your suggested change, so I've re-implemented it manually.  (done)

Global data flow
----------------

Global data flow tracks data flow throughout the entire program, and is therefore more powerful than local data flow.
However, global data flow is less precise than local data flow, and the analysis typically requires significantly more time and memory to perform.

.. pull-quote:: Note

>    .. include:: ../reusables/path-problem.rst

Reviewer: This doesn't render anything in the markdown viewer on GitHub, but I assume it resolves correctly once we put it on the microsite?

Author: It turns out you can get a slightly more accurate rendering from the "Docs - Generate Sphinx" CI check (download the `artifact.zip`), though there's a bit more effort involved.  The linked resource renders there as:
```
Note

You can model data flow paths in CodeQL by creating path queries. To view data flow paths generated by a path query in CodeQL for VS Code, you need to make sure that it has the correct metadata and select clause. For more information, see Creating path queries.
```

Reviewer 2: If you download the artefact from the "Docs Sphinx" check run, you can explore the HTML versions of these files. In this case, it looks like this:

<img width="732" alt="image" src="https://github.com/github/codeql/assets/1877141/fa52a280-6b19-49df-a547-26bd8d915b39">

Using global data flow
~~~~~~~~~~~~~~~~~~~~~~

You can use the global data flow library by implementing the module ``DataFlow::ConfigSig``:

.. code-block:: ql

   import codeql.swift.dataflow.DataFlow

   module MyDataFlowConfiguration implements DataFlow::ConfigSig {
     predicate isSource(DataFlow::Node source) {
       ...
     }

     predicate isSink(DataFlow::Node sink) {
       ...
     }
   }

   module MyDataFlow = DataFlow::Global<MyDataFlowConfiguration>;

These predicates are defined in the configuration:

-  ``isSource`` - defines where data may flow from.
-  ``isSink`` - defines where data may flow to.
-  ``isBarrier`` - optionally, restricts the data flow.
-  ``isAdditionalFlowStep`` - optionally, adds additional flow steps.

> The last line (``module MyDataFlow = ...``) performs data flow analysis using the configuration, and its results can be accessed with ``MyDataFlow::flow(DataFlow::Node source, DataFlow::Node sink)``:

Reviewer: I'm not sure I like the way this is phrased. The `module MyDataFlow = ...` declaration doesn't _perform_ the dataflow analysis. Rather, it instantiates the parameterized dataflow module with a concrete dataflow configuration so that you later can perform dataflow analysis. How about something like:

```suggestion
The last line (``module MyDataFlow = ...``) instantiates the parameterized module for data flow analysis by passing the configuration to the parameterized module. Data flow analysis can then be performed using ``MyDataFlow::flow(DataFlow::Node source, DataFlow::Node sink)``:
```

(maybe with a reference to https://codeql.github.com/docs/ql-language-reference/modules/#parameterized-modules somewhere?)

Author: Yeah, I was struggling with how to phrase this on the correct / intuitive scale, and quite possibly adding in a bit of my own misconception.  I'm happy with your improved phrasing.

  - Since this is a taint-tracking query, the ``TaintTracking::Global`` module is used.
  - The ``isSource`` predicate defines sources as any ``StringLiteralExpr``.
  - The ``isSink`` predicate defines sinks as arguments to a ``CallExpr`` called "password".
  - The sources and sinks may need tuning to a particular use, for example if passwords are represented by a type other than ``String`` or passed in arguments of a different name than "password".

.. code-block:: ql

   import swift
   import codeql.swift.dataflow.DataFlow
   import codeql.swift.dataflow.TaintTracking

   module ConstantPasswordConfig implements DataFlow::ConfigSig {
     predicate isSource(DataFlow::Node node) { node.asExpr() instanceof StringLiteralExpr }

     predicate isSink(DataFlow::Node node) {
       // any argument called `password`
       exists(CallExpr call | call.getArgumentWithLabel("password").getExpr() = node.asExpr())
     }

   module ConstantPasswordFlow = TaintTracking::Global<ConstantPasswordConfig>;

   from DataFlow::Node sourceNode, DataFlow::Node sinkNode
   where ConstantPasswordFlow::flow(sourceNode, sinkNode)

>    select sinkNode, "The value '" + sourceNode.toString() + "' is used as a constant password."

Reviewer: This would give the user clickable links to the source. Would that be better?

```suggestion
   select sinkNode, "The value $@ is used as a constant password.", sourceNode, sourceNode.toString()
```

The following global taint-tracking query finds places where a value from a remote or local user input is used as an argument to the SQLite ``Connection.execute(_:)`` function.
  - Since this is a taint-tracking query, the ``TaintTracking::Global`` module is used.
  - The ``isSource`` predicate defines sources as a ``FlowSource`` (remote or local user input).
  - The ``isSink`` predicate defines sinks as the first argument in any call to ``Connection.execute(_:)``.

.. code-block:: ql

   import swift
   import codeql.swift.dataflow.DataFlow
   import codeql.swift.dataflow.TaintTracking
   import codeql.swift.dataflow.FlowSources

   module SqlInjectionConfig implements DataFlow::ConfigSig {
     predicate isSource(DataFlow::Node node) { node instanceof FlowSource }

     predicate isSink(DataFlow::Node node) {
       exists(CallExpr call |

>          call.getStaticTarget().(MethodDecl).hasQualifiedName("Connection", "execute(_:)") and

Reviewer: `MethodDecl` has been renamed to `Method`
```suggestion
         call.getStaticTarget().(Method).hasQualifiedName("Connection", "execute(_:)") and
```

The following sections describe how to use the libraries for local data flow, global data flow, and taint tracking.
For a more general introduction to modeling data flow, see ":ref:`About data flow analysis <about-data-flow-analysis>`."

Local data flow
---------------

Local data flow tracks the flow of data within a single function. Local data flow is easier, faster, and more precise than global data flow. Before looking at more complex tracking, you should always consider local tracking because it is sufficient for many queries.

Using local data flow
~~~~~~~~~~~~~~~~~~~~~

You can use the local data flow library by importing the ``DataFlow`` module. The library uses the class ``Node`` to represent any element through which data can flow.
The ``Node`` class has a number of useful subclasses, such as ``ExprNode`` for expressions and ``ParameterNode`` for parameters. You can map between data flow nodes and expressions/control-flow nodes using the member predicates ``asExpr`` and ``getCfgNode``:

.. code-block:: ql

     class Node {
       /**

>         * Gets this node's underlying expression, if any.

Reviewer 2: Suggest simplifying, similar to Ruby:

```suggestion
        * Gets the expression that corresponds to this node, if any.
```

Author: Thanks.

Note to self: I should follow up with changes to the library code these snippets are quoted from as well.

        */
       Expr asExpr() { ... }

       /**

>         * Gets this data flow node's corresponding control flow node.

Reviewer 2: ```suggestion
        * Gets the control flow node that corresponds to this data flow node. 
```

Global data flow
----------------

Global data flow tracks data flow throughout the entire program, and is therefore more powerful than local data flow.
However, global data flow is less precise than local data flow, and the analysis typically requires significantly more time and memory to perform.

.. pull-quote:: Note

   .. include:: ../reusables/path-problem.rst

Using global data flow
~~~~~~~~~~~~~~~~~~~~~~

You can use the global data flow library by implementing the module ``DataFlow::ConfigSig``:

.. code-block:: ql

   import codeql.swift.dataflow.DataFlow

   module MyDataFlowConfiguration implements DataFlow::ConfigSig {
     predicate isSource(DataFlow::Node source) {
       ...
     }

     predicate isSink(DataFlow::Node sink) {
       ...
     }
   }

   module MyDataFlow = DataFlow::Global<MyDataFlowConfiguration>;

These predicates are defined in the configuration:

-  ``isSource`` - defines where data may flow from.
-  ``isSink`` - defines where data may flow to.
-  ``isBarrier`` - optionally, restricts the data flow.
-  ``isAdditionalFlowStep`` - optionally, adds additional flow steps.

> The last line (``module MyDataFlow = ...``) instantiates the parameterized module for data flow analysis by passing the configuration to the parameterized module. Data flow analysis can then be performed using ``MyDataFlow::flow(DataFlow::Node source, DataFlow::Node sink)``:

Reviewer 2: I think we can simplify this to something like:

```suggestion
The last line (``module MyDataFlow = ...``) passes the configuration to the parameterized module to instantiate the parameterized module ready for data flow analysis. Data flow analysis can then be performed using ``MyDataFlow::flow(DataFlow::Node source, DataFlow::Node sink)``:
```

Author: This is the change I'm least sure about, since "instantiates the parameterized module" really is the thing that line of code is doing, whereas "passes the configuration" is an important but secondary detail.

@MathiasVP I believe you commented on this bit previously - do you have an opinion?

Reviewer: Hm. I agree with @geoffw0. I don't think this change makes the sentence more clear. I'd much rather keep the phrasing as-is (if @felicitymay don't mind).

Reviewer 2: Thanks. Resolving this comment.

.. code-block:: ql

   from DataFlow::Node source, DataFlow::Node sink
   where MyDataFlow::flow(source, sink)
   select source, "Dataflow to $@.", sink, sink.toString()

Using global taint tracking
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Global taint tracking is to global data flow what local taint tracking is to local data flow.
That is, global taint tracking extends global data flow with additional non-value-preserving steps.

> The global taint tracking library uses the same configuration module as the global data flow library but taint flow analysis is performed with ``TaintTracking::Global``:

Reviewer 2: ```suggestion
The global taint tracking library uses the same configuration module as the global data flow library. You can perform taint flow analysis using ``TaintTracking::Global``:
```

.. code-block:: ql

   module MyTaintFlow = TaintTracking::Global<MyDataFlowConfiguration>;

   from DataFlow::Node source, DataFlow::Node sink
   where MyTaintFlow::flow(source, sink)
   select source, "Taint flow to $@.", sink, sink.toString()

Predefined sources
~~~~~~~~~~~~~~~~~~

> The data flow library module ``codeql.swift.dataflow.FlowSources`` contains a number of predefined sources, providing a good starting point for defining data flow and taint flow based security queries.

Reviewer 2: ```suggestion
The data flow library module ``codeql.swift.dataflow.FlowSources`` contains a number of predefined sources that you can use to write security queries to track data flow and taint flow.
```

-  The class ``RemoteFlowSource`` represents data flow from remote network inputs and from other applications.
-  The class ``LocalFlowSource`` represents data flow from local user input.
-  The class ``FlowSource`` includes both of the above.

Examples of global data flow
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following global taint-tracking query finds places where a string literal is used in a function call argument named "password".
  - Since this is a taint-tracking query, the ``TaintTracking::Global`` module is used.
  - The ``isSource`` predicate defines sources as any ``StringLiteralExpr``.
  - The ``isSink`` predicate defines sinks as arguments to a ``CallExpr`` called "password".

>   - The sources and sinks may need tuning to a particular use, for example if passwords are represented by a type other than ``String`` or passed in arguments of a different name than "password".

Reviewer 2: ```suggestion
  - The sources and sinks may need tuning to a particular use, for example, if passwords are represented by a type other than ``String`` or passed in arguments of a different name than "password".
```
